web/api/rest-api-design/Error-Response-Strategy.md

```diff
@@ -1,44 +1,97 @@
+
+For e.g. The types of responses are grouped into the following top level categories.
 
-* You should always return consistent errors. This way your error
-  handling code on the front end can be somewhat generic, instead of
-  having to adapt to different response types based on the API.
+* `2xx` **Success** - This category indicates the action requested by the client was received, understood, and accepted.
+* `4xx` **Client Error** - This category indicate that the error seems to have been caused by the client.
+* `5xx` **Server Error** - This category indicates the server failed to fulfill a request.
 
-* This is an argument in favor of returning a list of error objects,
-  rather than a single error object, so that the front end doesn’t
-  have to check every time what the shape of the error response is
+**Importance of the Client vs. Server Error Distinction** - `500` server errors signal that something went wrong on the backend *that shouldn't have*. Most known errors should be a `4xx`. Most `500` errors are bugs. Keep this in mind as you select which type of error response your situation calls for.
 
-### Example of error strategies for each type of error (i.e. server, client, validation, etc.)
+Since there are a lot of status codes, it's advisable to pick a well-known subset of HTTP responses and return mainly those.
 
-#### Errors
+* `200` **OK** - Response to a successful GET, PUT, PATCH or DELETE. Can also be used for a POST that doesn't result in a creation.
+* `201` **Created** - Response to a POST that results in a creation.
+* `404` **Not Found** - When a non-existent resource is requested.
+* `422` **Unprocessable Entity** - Used for validation errors.
+* `409` **Conflict** - When we cannot process the request due to the current state of the server.
+* `412` **Precondition Failed** - A precondition like [optimistic locking](./Concurrency-Control.md) has failed.
 
-Any logic or presentation of the error in the front end should be
-based on the code here, never match against the message. Some people
-use `ints` for this but a short string is more readable:
+### Use a constant error code to uniquely identify errors
 
-* 2123 vs. “MISSING_PARAMETER”
+A constant error code in the body allows the logic on the frontend to understand and handle the error correctly. Note this is different from the HTTP status code.
 
-#### Messages
+Any logic that handles this error code should match on the error code and not on an error string. This ensures error handling won’t break if the message changes.
+
+For the error code itself, consider using an enumerated short string for readability. This provides the twofold benefit of being a constant value, but also being human-readable.
+
+```json
+  "Code": "MISSING_PARAMETER"
+```
 
-Messages are mainly used by developers who are reading a raw response,
-errors presented to users can be switched on the “code” parameter.
+### Provide Detailed Messages
```

Comment:
I have a couple of concerns around this suggestion, especially for any API that may be available on the *public* internet... 
Giving too much information "out the front door" can be a vector for security issues. So I believe there needs to be a balance struck between legibility to the API end-user and terseness... (Especially if the intended API end-users are "internal" folks, where ideally they would have access to seeing what errors are happening via the back-end, e.g. via logging.)

Reply:
Yup, makes sense, I've added that nuance to that section.